### PR TITLE
feat(runStaticServer): Serve a static directory easily

### LINKED ISSUE #379: Simple server API
In Python you can do this from a shell:

```
python -m http.server 5000
```

and this will start a simple static web server on 0.0.0.0:5000 that serves up the current directory as static files. We don't have a simple way to do this with httpuv, this is the equivalent invocation:

```R
httpuv::runServer(app = list(staticPaths=list("/"=httpuv::staticPath(getwd()))), host = "0.0.0.0", port = 5000)
```

It would be great to have a single function for this, with parameters for:
* host (default: 0.0.0.0 or 127.0.0.1?)
* port (default: hardcoded value)
* directory (default: `getwd()`)

OR, maybe we create a convenience function to create `list(staticPaths=list("/"=httpuv::staticPath(getwd())))` and that becomes the default value for the `app` parameter of `runServer` and `startServer`. Like `runServer <- function(host, port, app=staticSite("."), ...)`

My motivation for this is to help the story of distributing Shinylive/webR HTML/JS/CSS bundles to end users on Windows, where Python may not be installed. Those bundles cannot be run with the file: protocol, it must be http(s). I'd love to be able to tell them "you can start a little local web server with either `python -m http.server` or `R -e 'httpuv::runServer()'`".

cc @rpodcast @wch

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #379 

This PR adds a simple static server convenience function that serves as a user-friendly counter part to `runServer()`. It also includes a very simple, two-page site that can be used as an example or for testing purposes.

```r
pkgload::load_all()

website_dir <- system.file("example-static-site", package = "httpuv")
runStaticServer(dir = website_dir)
#> Serving: '/Users/garrick/work/rstudio/httpuv/inst/example-static-site'
#> View at: http://127.0.0.1:7446
#> Browsing http://127.0.0.1:7446
#> Press Esc or Ctrl + C to stop the server
```

### Running in the foreground (default)

By default, `runStaticServer()` runs a blocking server on port `7446` (`shin(y)`). If the port is in use, it will find another port. It prints consistent messages for the the user, always showing the directory being served and the full address. If `browse` is `TRUE` (the default in interactive sessions), `runStaticServer()` will call `browseURL()` to open the served pages. When run in the foreground, it also prints shutdown instructions.

I went back and forth on including both the `View at: http://...` and `Browsing http://...` messages, but decided to keep both. `utils::browseURL()` prints the `Browsing` message, but I wanted the `View at:` line to be stable in console output in case the output is parsed by a parent process.

### `background = TRUE`

When `background = TRUE`, `runStaticServer()` will start a background server and returns the `server` object invisibly. Users can call `stopServer(server)` or `stopAllServers()` to stop the running server.

### `randomPort()` with a single port

I updated `randomPort()` so that it can be used to test a single port value. Today I rediscovered that `sample(10, 1)` doesn't return `10` but instead returns a single value sampled from the integers `1:10`

``` r
sample(10, 1)
#> [1] 5
sample(10, 1)
#> [1] 4
sample(10, 1)
#> [1] 6
```

## TODO

- [x] Document `browse` param
- [x] Feedback on API
- [x] Add tests

### CODE REVIEW

File: R/staticServer.R [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+requested_or_random_port <- function(port, host = "127.0.0.1") {
+  tryCatch(
+    # Is the requested port open?
+    randomPort(port, port, host = host),
+    error = function(err) {
+      msg <- conditionMessage(err)
+      if (!grepl("Cannot find an available port", msg, fixed = TRUE)) {
+        stop(msg)
```
Comment (jcheng5) [inline R/staticServer.R:70]: If the intention is to re-raise the error, then how about:
```suggestion
        stop(err)
```

File: R/staticServer.R [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  browse = interactive()
+) {
+  stopifnot(
+    "`port` must be an integer" = is.numeric(port),
+    "`port` must be an integer" = port == as.integer(port),
+    "`port` must be a single integer" = length(port) == 1
+  )
+
+  browse <- isTRUE(browse)
+  port <- requested_or_random_port(port, host)
```
Comment (jcheng5) [inline R/staticServer.R:40]: I don't love it that a user can explicitly ask for a certain port and get a different one back. I'd prefer either just failing if the port is in use (even if it's the default port), or at most, only falling back to a random port if `missing(port)`.

Comment (gadenbuie) [reply R/staticServer.R:40]: Agreed, this bugged me too. I'll use a different pattern where we'll use a random port only if the default port is unavaiable and the user didn't ask for a specific port.

File: R/staticServer.R (hunk 1 above)
Comment (gadenbuie) [reply R/staticServer.R]: I refactored the internals of `randomPort()` to throw an error with a custom class (f5d699760ff01de7fde19f91a8e82fd353889a66) and then pulled out the try and throw logic into smaller helper functions in a637e085c88a1a95ce89386b3a33412c3e24e9b9.

Breaking apart the logic let me be more prescriptive about the default port handling and I ultimately didn't need the custom error class in this PR. But I've run into situations where a classed error would be helpful and think it's worth keeping

File: R/staticServer.R [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#'
+#' @export
+runStaticServer <- function(
+  dir = getwd(),
+  host = "127.0.0.1",
+  port = NULL,
+  background = FALSE,
+  browse = interactive()
+) {
+  root <- staticPath(dir)
```
Comment (gadenbuie) [inline R/staticServer.R:40]: Add dots to pass options down to `staticPath()`

```suggestion
  dir = getwd(),
  host = "127.0.0.1",
  port = NULL,
  ...,
  background = FALSE,
  browse = interactive()
) {
  root <- staticPath(dir, ...)
```

Comment (gadenbuie) [reply R/staticServer.R]: Done in ed42fdfeedf8c97be4bcaa1d42bd86b260ce3dc8

File: R/staticServer.R [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  if (is.null(port)) {
+    port <- if (is_port_available(7446, host)) 7446 else randomPort(host = host)
+  } else {
+    stopifnot(
+      "`port` must be an integer" = is.numeric(port),
+      "`port` must be an integer" = port == as.integer(port),
+      "`port` must be a single integer" = length(port) == 1,
+      "`port` must be a positive integer" = port > 0,
+      "`port` must be less than 65536" = port < 65536
```
Comment (jcheng5) [inline R/staticServer.R:52]: Wow, TIL stopifnot could do this

**Review by jcheng5 — APPROVED** [review]

**Comment by jcheng5** [comment]
Just a NEWS.md update and then I think this is good to go! Thanks for getting on this so quickly.

File: R/staticServer.R (hunk 4 above)
Comment (gadenbuie) [reply R/staticServer.R:52]: Yeah! It was added in R 4.0, but it's backwards compatible in that if a condition isn't `TRUE` on R < 4.0 you'll get the typical `stopifnot(length(port) == 1)` message.

**Pull request merged**